src/shogun/classifier/Perceptron.cpp

```diff
@@ -77,42 +52,30 @@ bool CPerceptron::train_machine(CFeatures* data)
 		w.set_const(1.0 / num_feat);
 		bias=0;
 	}
-	else
-	{
-		w = get_w();
-	}
-	//loop till we either get everything classified right or reach max_iter
-	while (!converged && iter < max_iter)
+}
+
+void CPerceptron::iteration(CDotFeatures* f, CLabels* l)
+{
+	m_complete = true;
+	SGVector<float64_t> w = get_w();
+	SGVector<float64_t> output(f->get_num_vectors());
+	auto iter_train_labels = binary_labels(l)->get_int_labels().begin();
+	auto iter_output = output.begin();
+	for (const auto& v : DotIterator(f))
 	{
-		COMPUTATION_CONTROLLERS
-		converged=true;
-		auto iter_train_labels = train_labels.begin();
-		auto iter_output = output.begin();
-		for (const auto& v : DotIterator(features))
-		{
-			const auto true_label = *(iter_train_labels++);
-			auto& predicted_label = *(iter_output++);
+		const auto true_label = *(iter_train_labels++);
+		auto& predicted_label = *(iter_output++);
 
-			predicted_label = v.dot(w) + bias;
+		predicted_label = v.dot(w) + bias;
 
-			if (CMath::sign<float64_t>(predicted_label) != true_label)
-			{
-				converged = false;
-				const auto gradient = learn_rate * true_label;
-				bias += gradient;
-				v.add(gradient, w);
-			}
+		if (CMath::sign<float64_t>(predicted_label) != true_label)
+		{
+			m_complete = false;
+			const auto gradient = learn_rate * true_label;
+			bias += gradient;
+			v.add(gradient, w);
 		}
-
-		iter++;
 	}
-
```

Comment:
for error checking after loop ops maybe we should add an end_training() method to IterativeMachine